lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+# (c) 2019 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os.path
+import pkgutil
+import re
+import sys
+
+from importlib import import_module
+from types import ModuleType
+
+from ansible import constants as C
+from ansible.module_utils._text import to_native
+from ansible.module_utils.six import iteritems, string_types
+
+_SYNTHETIC_PACKAGES = {
+    'ansible_collections.ansible': dict(type='pkg_only'),
+    'ansible_collections.ansible.core': dict(type='pkg_only'),
+    'ansible_collections.ansible.core.plugins': dict(type='map', map='ansible.plugins'),
+    'ansible_collections.ansible.core.plugins.module_utils': dict(type='map', map='ansible.module_utils', graft=True),
+    'ansible_collections.ansible.core.plugins.modules': dict(type='flatmap', flatmap='ansible.modules', graft=True),
+}
+
+# TODO: tighten this up to subset Python identifier requirements
+_collection_role_name_re = re.compile(r'^(\w+)\.(\w+)\.(\w+)$')
+
+# FIXME: exception handling/error logging
+class AnsibleCollectionLoader(object):
+    def __init__(self):
+        self._configured_paths = C.config.get_config_value('INSTALLED_CONTENT_ROOTS')
+
+        if isinstance(self._configured_paths, string_types):
+            self._configured_paths = [self._configured_paths]
+        elif self._configured_paths is None:
+            self._configured_paths = []
+
+        # expand any placeholders in configured paths
+        self._configured_paths = [os.path.expanduser(p) for p in self._configured_paths]
```

Comment:
Wondering if it would make sense to change AnsibleCollectionLoader so that it only has to know about one content root (ContentRootLoader?).  And adding a chain of responsibility style layer above (I'll call it ContentRootLoaderChain for now).  ContentRootLoaderChain would have-a ordered list of ContentRootLoader's (one for each content root), and ContentRootLoaderChain would be responsible for searching until it finds what it needs.

In addition to the ContentRootLoader instance for each installed_content_root, the ContentRootLoaderChain
could add additional links to check. For example, maybe a cache layer. Could also have a ContentRootLoader instance for each of 'ansible.legacy' and 'ansible.core' that is added to the end of the chain.

Logic for scoping/priority/precedence  could potentially be mostly located in ContentRootLoaderChain.

I think AnsibleContentLoader / ContentRootLoader would be simpler if they only had to know about one content root.

Reply:
Oh, and the playbook adjacent collection loader logic could be extracted to it's own class (PlaybookContentRootLoader?) and added to the ContentRootLoaderChain